sql/src/main/java/org/apache/druid/sql/calcite/schema/MetadataSegmentView.java

```diff
@@ -0,0 +1,255 @@
+    publishedSegments.entrySet().removeIf(e -> e.getValue() != timestamp);
+    cachePopulated.set(true);
+  }
+
+  public Iterator<DataSegment> getPublishedSegments()
+  {
+    if (isCacheEnabled) {
+      Preconditions.checkState(
+          lifecycleLock.awaitStarted(1, TimeUnit.MILLISECONDS) && cachePopulated.get(),
+          "hold on, still syncing published segments"
+      );
+      return publishedSegments.keySet().iterator();
+    } else {
+      return getMetadataSegments(
+          coordinatorDruidLeaderClient,
+          jsonMapper,
+          responseHandler,
+          segmentWatcherConfig.getWatchedDataSources()
+      );
+    }
+  }
+
+  // Note that coordinator must be up to get segments
+  private JsonParserIterator<DataSegment> getMetadataSegments(
+      DruidLeaderClient coordinatorClient,
+      ObjectMapper jsonMapper,
+      BytesAccumulatingResponseHandler responseHandler,
+      Set<String> watchedDataSources
+  )
+  {
+    String query = "/druid/coordinator/v1/metadata/segments";
+    if (watchedDataSources != null && !watchedDataSources.isEmpty()) {
+      log.debug(
+          "filtering datasources in published segments based on broker's watchedDataSources[%s]", watchedDataSources);
+      final StringBuilder sb = new StringBuilder();
+      for (String ds : watchedDataSources) {
+        sb.append("datasources=").append(ds).append("&");
+      }
+      sb.setLength(sb.length() - 1);
+      query = "/druid/coordinator/v1/metadata/segments?" + sb;
+    }
+    Request request;
+    try {
+      request = coordinatorClient.makeRequest(
+          HttpMethod.GET,
+          StringUtils.format(query),
+          false
+      );
+    }
+    catch (IOException e) {
+      throw new RuntimeException(e);
+    }
+    ListenableFuture<InputStream> future = coordinatorClient.goAsync(
+        request,
+        responseHandler
+    );
+
+    final JavaType typeRef = jsonMapper.getTypeFactory().constructType(new TypeReference<DataSegment>()
+    {
+    });
+    return new JsonParserIterator<>(
+        typeRef,
+        future,
+        request.getUrl().toString(),
+        null,
+        request.getUrl().getHost(),
+        jsonMapper,
+        responseHandler
+    );
+  }
+
+  private class PollTask implements Runnable
+  {
+    @Override
+    public void run()
+    {
+      long delayMS = pollPeriodInMS;
+      try {
+        final long pollStartTime = System.nanoTime();
+        poll();
+        final long pollEndTime = System.nanoTime();
+        final long pollTimeNS = pollEndTime - pollStartTime;
+        final long pollTimeMS = TimeUnit.NANOSECONDS.toMillis(pollTimeNS);
+        delayMS = Math.max(pollPeriodInMS - pollTimeMS, 0);
+      }
+      catch (Exception e) {
+        log.makeAlert(e, "Problem polling Coordinator.").emit();
+      }
+      finally {
+        if (!Thread.currentThread().isInterrupted()) {
+          scheduledExec.schedule(new PollTask(), delayMS, TimeUnit.MILLISECONDS);
```

Comment:
Why doesn't this class use `ScheduledExecutorService.sheduleAtFixedRate()` or `sheduleWithFixedDelay()`?

Reply:
i changed to use `schedule` because if `poll` is executed with fixed delay, there might be a possibility of two or more polls to be executed at the same time, if there's huge GC or `getMetadataSegments` takes longer than poll period.

Reply:
Both `scheduleAtFixedRate` and `sheduleWithFixedDelay` explicitly prohibit concurrent execution in their Javadocs. Also, this is not imaginably possible, because your `sheduledExec` has only a single thread.